packages/ai-chat/src/common/chat-model.ts

```diff
@@ -0,0 +1,753 @@
+
+import { Command, Emitter, Event, generateUuid, URI } from '@theia/core';
+import { MarkdownString, MarkdownStringImpl } from '@theia/core/lib/common/markdown-rendering';
+import { Position } from '@theia/core/shared/vscode-languageserver-protocol';
+import { ChatAgentLocation } from './chat-agents';
+import { ParsedChatRequest } from './chat-parsed-request';
+
+/**********************
+ * INTERFACES AND TYPE GUARDS
+ **********************/
+
+export type ChatChangeEvent =
+    | ChatAddRequestEvent
+    | ChatAddResponseEvent
+    | ChatRemoveRequestEvent;
+
+export interface ChatAddRequestEvent {
+    kind: 'addRequest';
+    request: ChatRequestModel;
+}
+
+export interface ChatAddResponseEvent {
+    kind: 'addResponse';
+    response: ChatResponseModel;
+}
+
+export type ChatRequestRemovalReason = 'removal' | 'resend' | 'adoption';
+
+export interface ChatRemoveRequestEvent {
+    kind: 'removeRequest';
+    requestId: string;
+    responseId?: string;
+    reason: ChatRequestRemovalReason;
+}
+
+export interface ChatModel {
+    readonly onDidChange: Event<ChatChangeEvent>;
+    readonly id: string;
+    readonly location: ChatAgentLocation;
+    getRequests(): ChatRequestModel[];
+    addRequest(parsedChatRequest: ParsedChatRequest, agentId?: string): ChatRequestModel;
+    isEmpty(): boolean;
+}
+
+export interface ChatRequest {
+    readonly text: string;
+    readonly displayText?: string;
+}
+
+export interface ChatRequestModel {
+    readonly id: string;
+    readonly session: ChatModel;
+    readonly request: ChatRequest;
+    readonly response: ChatResponseModel;
+    readonly message: ParsedChatRequest;
+    readonly agentId?: string;
+}
+
+export interface ChatProgressMessage {
+    kind: 'progressMessage';
+    id: string;
+    status: 'inProgress' | 'completed' | 'failed';
+    content: string;
+}
+
+export interface BaseChatResponseContent {
+    kind: string;
+    /**
+     * Represents the content as a string. Returns `undefined` if the content
+     * is purely informational and/or visual and should not be included in the overall
+     * representation of the response.
+     */
+    asString?(): string | undefined;
+    merge?(nextChatResponseContent: BaseChatResponseContent): boolean;
+}
+
+export const isBaseChatResponseContent = (
+    obj: unknown
+): obj is BaseChatResponseContent =>
+    !!(
+        obj &&
+        typeof obj === 'object' &&
+        'kind' in obj &&
+        typeof (obj as { kind: unknown }).kind === 'string'
+    );
+
+export const hasAsString = (
+    obj: BaseChatResponseContent
+): obj is Required<Pick<BaseChatResponseContent, 'asString'>> &
+BaseChatResponseContent => obj.asString !== undefined;
+
+export const hasMerge = (
+    obj: BaseChatResponseContent
+): obj is Required<Pick<BaseChatResponseContent, 'merge'>> &
+BaseChatResponseContent => obj.merge !== undefined;
+
+export interface TextChatResponseContent
+    extends Required<BaseChatResponseContent> {
```

Comment:
Not sure I can follow. The architecture of the response contents is that:
 - `ChatAgent`s can return arbitrary responses of all `kind`s they like
 - We don't want to restrict which `kind` of content Theia supports, but we want to generically handle them
 - A typical extension use case is that a contributed agent returns all kind of specialized responses. The same extension registers renderers for these new `kind` and thereby Theia is able to render them.
 - However if the contribution plays nice, it can opt-in to some functionality so we can handle them better generically
 
Therefore there is a:
 - mandatory `kind` of type `string` so that the content can be identified
 - two optional functionalities, i.e. `asString?` and `merge?`.

In our own types we definitely opt-in so we reused the base interface with a `Required`. How would you like to see this differently?

For `kind` we used this mechanism instead of a union of predefined values as we expect there to be custom `kind`s. Using a specialized `kind` in each interface let's Typescript automatically conclude that its sibling properties must exist. For example checking for kind `code` let's Typescript infer that there is a mandatory `code` string, and optional `language` and `location` props.

Reply:
But that makes no sense: we never call `merge` polymorphically, so we need to know the concrete subtype anyway. Then why does it make sense to have an abstract method? We could name the method differently in every subtype and the code would be the same.
Also, the set of response types is not open to extension, you explicitly define it as 
```
export type ChatResponseContent =
    | BaseChatResponseContent
    | TextChatResponseContent
...
```
There really should be a registry of response types where we can register mergers and renderers per response `kind`.

Reply:
We have a generic merge functionality already in place supporting merge for all existing kinds and all contributed kinds: We are calling `merge` when the `kind` string is the same on the last and the incoming element

In `BaseChatResponseContent` the `type` is just defined as a string. So `merge` will be called for all sub types, including custom ones.

See this code
```ts
  if (lastElement?.kind === nextContent.kind && BaseChatResponseContent.hasMerge(lastElement)) {
      const mergeSuccess = lastElement.merge(nextContent);
      if (!mergeSuccess) {
          this._content.push(nextContent);
      }
  } else {
      this._content.push(nextContent);
  }
```

So if a user adds their own `NiceGraphicalChatResponseContent` and defines `kind: 'niceGraphicalResponseContent'` with a `merge` function, then the `merge` of this custom kind will be called.

---

However, I agree that adding a more powerful merge technology would make sense, as we also indicated in the `TODO`

> // TODO: Support more complex merges affecting different content than the last, e.g. via some kind of ProcessorRegistry

A registry in which we can not only register renderers but also mergers makes sense to me, so we could refactor the code here. Likely we want two different registries?

Reply:
I added it to the follow up: https://github.com/eclipse-theia/theia/issues/14143